File: week12/3/largest_number.c
#include <stdio.h>
#include <stdlib.h> // qsort
#include <string.h>

// helper function: compare strings based on custom sorting rule
// quicksort-like approach
/*
const pointers: a value cannot be modified
read-only access in a comparator function, commonly used in qsort
*/
int quickCompare(const void* a, const void* b) {
    // comparator function for string converted array
    // void*: pointer to data of an unspecified type
    // char**: pointer to a char* (pointing to the string stored in char*)
    // *(char**)a
    // : (char**)a casts the generic void* to a char** (pointer to a pointer to a char)
    // : * in front dereferences it, giving the char* (a pointer to the actual string)
    // extracted char* (string pointer) assigned to x
    char* x = *(char**)a;
    char* y = *(char**)b;

    // concatenate strings in both possible orders
    char xy[strlen(x) + strlen(y) + 1];
    char yx[strlen(y) + strlen(x) + 1];
    sprintf(xy, "%s%s", x, y);
    sprintf(yx, "%s%s", y, x);
    // lexicographically compare yx with xy
    // strcmp(yx, xy) < 0 : yx < xy
    // strcmp(yx, xy) > 0 : yx > xy
    // largest concatenation comes first in sorted array (reverse order for descending sort)
    return strcmp(yx, xy);
}


// Main function: form the largest number out of nums array
char* largestNumber(int* nums, int numsSize) {
    // convert element of nums from integer to strings
    char** str_nums = malloc(numsSize * sizeof(char*));
    for (int i = 0; i < numsSize; i++) {
        str_nums[i] = malloc(12 * sizeof(char)); // enough to hold any 32-bit integer
        sprintf(str_nums[i], "%d", nums[i]);
    } // str_nums formed!

    // void qsort( void* ptr, size_t count, size_t size, int (*comp)(const void*, const void*) );
    // sort str_nums using custom comparator
    qsort(str_nums, numsSize, sizeof(char*), quickCompare);

    // handle edge case: largest number is 0
    if (strcmp(str_nums[0], "0") == 0) {
        // free memory
        for (int i = 0; i < numsSize; i++) free(str_nums[i]);
        free(str_nums);
        // return "0"
        char* result = malloc(2 * sizeof(char));
        strcpy(result, "0");
        return result;
    }

    // concatenate sorted string numbers into the largest number
    int ans_length = 0;
    for (int i=0; i < numsSize; i++) {
        ans_length += strlen(str_nums[i]);
    }
    // memory allocation for ans
    char* ans = malloc((ans_length + 1) * sizeof(char));
    ans[0] = '\0'; // initialize as empty string
    for (int i = 0; i < numsSize; i++) {
        strcat(ans, str_nums[i]);
        free(str_nums[i]);
    }
    // free memory
    free(str_nums);

    return ans;
}

// example usage
int main() {
    int nums1[] = {10, 2};
    printf("%s\n", largestNumber(nums1, 2)); // Output: "210"

    int nums2[] = {3, 30, 34, 5, 9};
    printf("%s\n", largestNumber(nums2, 5)); // Output: "9534330"

    return 0;
}

File: week12/3/largest_number.py
# Question:
# Given a list of non-negative integers nums, arrange them such that they form the largest number and return it.
# Since the result may be very large, you need to return it as a string instead of an integer.

# Example 1:
# Input: nums = [10, 2]
# Output: "210"

# Example 2:
# Input: nums = [3, 30, 34, 5, 9]
# Output: "9534330"

# Constraints:
# 1 <= nums.length <= 100
# 0 <= nums[i] <= 10^9

def largestNumber(nums):
    # convert all numbers from int to string
    str_nums = []
    for num in nums:
        str_nums.append(str(num))

    # recursive sorting algorithm: quicksort-like approach
    # sort string converted array
    def custom_sort(str_arr):

        # base case: already sorted (same or less than one element)
        if len(str_arr) <= 1:
            return str_arr
        
        # choose the first element as pivot
        pivot = str_arr[0]
        left = [] # elements smaller than pivot
        right = [] # elements larger than pivot

        # partition array based on custom comparison
        for i in range(1, len(str_arr)):
            # string for placement str_arr[i]: start from second element
            # for each element in the array (except the pivot),
            # compare it with the pivot by concatenating strings in both possible orders
            # e.g. comparison:  934  v.s.  349
            # print("comparison: ", str_arr[i] + pivot, " v.s. ", pivot + str_arr[i])
            if str_arr[i] + pivot > pivot + str_arr[i]:
                # str_arr[i] should come before pivot
                left.append(str_arr[i])
            else:
                # # str_arr[i] should come after pivot
                right.append(str_arr[i])
        
        # recursively sort the left and right partitions
        # then combine with the pivot
        # print("current sorted: ", custom_sort(left) + [pivot] + custom_sort(right))
        return custom_sort(left) + [pivot] + custom_sort(right)
    
    # sort array of num strings
    sorted_nums = custom_sort(str_nums)

    # concatenate sorted numbers into the largest number
    ans = "".join(sorted_nums)

    # handle edge case: largest number is 0
    if ans[0] != "0":
        return ans
    else:
        return "0"
    

# example 1
nums1 = [10, 2]
print(largestNumber(nums1)) # "210"

# example 2
nums2 = [3,30,34,5,9]
print(largestNumber(nums2)) # "9534330"